pkg/controller/machineset.go

```diff
@@ -493,35 +493,37 @@ func (c *controller) reconcileClusterMachineSet(key string) error {
 		return err
 	}
 
-	if machineSet.DeletionTimestamp != nil {
-		if finalizers := sets.NewString(machineSet.Finalizers...); !finalizers.Has(DeleteFinalizerName) {
-			return nil
-		}
-		if len(filteredMachines) == 0 {
-			c.deleteMachineSetFinalizers(machineSet)
-			return nil
-		}
-		glog.V(4).Infof("Deleting all child machines as MachineSet %s has set deletionTimestamp", machineSet.Name)
-		c.terminateMachines(filteredMachines, machineSet)
-		return nil
-	}
-
 	machineSetNeedsSync := c.expectations.SatisfiedExpectations(key)
 	var manageReplicasErr error
-	if machineSetNeedsSync {
+
+	if machineSetNeedsSync && machineSet.DeletionTimestamp == nil {
+		// manageReplicas is the core machineSet method where scale up/down occurs
+		// It is not called when deletion timestamp is set
 		manageReplicasErr = c.manageReplicas(filteredMachines, machineSet)
+
+	} else if machineSet.DeletionTimestamp != nil {
+		// When machineSet if triggered for deletion
+
+		if len(filteredMachines) == 0 {
+			// If machines backing a machineSet are zero,
+			// remove the machineSetFinalizer
+			c.deleteMachineSetFinalizers(machineSet)
+		} else if finalizers := sets.NewString(machineSet.Finalizers...); finalizers.Has(DeleteFinalizerName) {
+			// Trigger deletion of machines backing the machineSet
+			glog.V(4).Infof("Deleting all child machines as MachineSet %s has set deletionTimestamp", machineSet.Name)
+			c.terminateMachines(filteredMachines, machineSet)
+		}
 	}
-	//glog.V(2).Infof("Print manageReplicasErr: %v ",manageReplicasErr) //Remove
 
 	machineSet = machineSet.DeepCopy()
 	newStatus := calculateMachineSetStatus(machineSet, filteredMachines, manageReplicasErr)
 
 	// Always updates status as machines come up or die.
 	updatedMachineSet, err := updateMachineSetStatus(c.controlMachineClient, machineSet, newStatus)
-	if err != nil {
+	if err != nil && !apierrors.IsNotFound(err) {
```

Comment:
It will be good to log the error `IsNotFound{}` even if we don't want to return it.

Reply:
This log occurs often (object not found), and logging this would flood the logs with unnecessary statements?

Reply:
Then may be in some higher log level? so that it will not be printed by default.